common/cuda_hip/solver/batch_cg_kernels.hpp.inc

```diff
@@ -135,13 +135,45 @@ __global__ void apply_kernel(const int max_iter,
     auto warp_grp = group::tiled_partition<tile_size>(thread_block);
 
     for (size_type ibatch = blockIdx.x; ibatch < nbatch; ibatch += gridDim.x) {
+        const int gmem_offset =
+            ibatch * sconf.gmem_stride_bytes / sizeof(ValueType);
         extern __shared__ char local_mem_sh[];
-        ValueType* const r_sh = reinterpret_cast<ValueType*>(local_mem_sh);
-        ValueType* const z_sh = r_sh + nrows;
-        ValueType* const p_sh = z_sh + nrows;
-        ValueType* const Ap_sh = p_sh + nrows;
-        ValueType* const x_sh = Ap_sh + nrows;
-        ValueType* const prec_work_sh = x_sh + nrows;
+        ValueType* r_sh;
+        ValueType* z_sh;
+        ValueType* p_sh;
+        ValueType* Ap_sh;
+        ValueType* x_sh;
+        ValueType* prec_work_sh;
+        if (sconf.n_shared >= 1) {
+            r_sh = reinterpret_cast<ValueType*>(local_mem_sh);
+        } else {
+            r_sh = workspace + gmem_offset;
+        }
+        if (sconf.n_shared == 1) {
+            z_sh = workspace + gmem_offset;
+        } else {
+            z_sh = r_sh + padded_vec_len;
+        }
+        if (sconf.n_shared == 2) {
+            p_sh = workspace + gmem_offset;
+        } else {
+            p_sh = z_sh + padded_vec_len;
+        }
+        if (sconf.n_shared == 3) {
+            Ap_sh = workspace + gmem_offset;
+        } else {
+            Ap_sh = p_sh + padded_vec_len;
+        }
+        if (!sconf.prec_shared && sconf.n_shared == 4) {
+            prec_work_sh = workspace + gmem_offset;
+        } else {
+            prec_work_sh = Ap_sh + padded_vec_len;
```

Comment:
Don't you need `prec_storage_bytes` somehow here instead?

Reply:
Its included and used to compute aligned storage in the `gmem_offset` variable